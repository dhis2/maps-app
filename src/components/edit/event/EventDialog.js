import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import i18n from '@dhis2/d2-i18n';
import { NoticeBox } from '@dhis2/ui';
import {
    Tab,
    Tabs,
    NumberField,
    Checkbox,
    ImageSelect,
    ColorPicker,
} from '../../core';
import ProgramSelect from '../../program/ProgramSelect';
import ProgramStageSelect from '../../program/ProgramStageSelect';
import EventStatusSelect from './EventStatusSelect';
import RelativePeriodSelect from '../../periods/RelativePeriodSelect';
import StartEndDates from '../../periods/StartEndDates';
import FilterGroup from '../../filter/FilterGroup';
import StyleByDataItem from '../../dataItem/StyleByDataItem';
import CoordinateField from '../../dataItem/CoordinateField';
import OrgUnitTree from '../../orgunits/OrgUnitTree';
import UserOrgUnitsSelect from '../../orgunits/UserOrgUnitsSelect';
import SelectedOrgUnits from '../../orgunits/SelectedOrgUnits';
import {
    EVENT_COLOR,
    EVENT_RADIUS,
    EVENT_BUFFER,
} from '../../../constants/layers';
import styles from '../styles/LayerDialog.module.css';

import {
    setProgram,
    setProgramStage,
    setEventStatus,
    setEventCoordinateField,
    setEventClustering,
    setEventPointColor,
    setEventPointRadius,
    setOrgUnitRoot,
    setUserOrgUnits,
    toggleOrgUnit,
    setPeriod,
    setAreaRadius,
} from '../../../actions/layerEdit';

import {
    getPeriodFromFilters,
    getOrgUnitsFromRows,
    getOrgUnitNodesFromRows,
    getUserOrgUnitsFromRows,
} from '../../../util/analytics';
import { getStartEndDateError } from '../../../util/time';
import { cssColor } from '../../../util/colors';

export class EventDialog extends Component {
    static propTypes = {
        areaRadius: PropTypes.number,
        columns: PropTypes.array,
        defaultPeriod: PropTypes.string,
        endDate: PropTypes.string,
        eventClustering: PropTypes.bool,
        eventStatus: PropTypes.string,
        eventCoordinateField: PropTypes.string,
        eventPointColor: PropTypes.string,
        eventPointRadius: PropTypes.number,
        filters: PropTypes.array,
        onLayerValidation: PropTypes.func.isRequired,
        program: PropTypes.shape({
            id: PropTypes.string.isRequired,
        }),
        programStage: PropTypes.shape({
            id: PropTypes.string.isRequired,
        }),
        rows: PropTypes.array,
        startDate: PropTypes.string,
        setProgram: PropTypes.func.isRequired,
        setProgramStage: PropTypes.func.isRequired,
        setEventStatus: PropTypes.func.isRequired,
        setEventCoordinateField: PropTypes.func.isRequired,
        setEventClustering: PropTypes.func.isRequired,
        setEventPointColor: PropTypes.func.isRequired,
        setEventPointRadius: PropTypes.func.isRequired,
        setOrgUnitRoot: PropTypes.func.isRequired,
        setUserOrgUnits: PropTypes.func.isRequired,
        toggleOrgUnit: PropTypes.func.isRequired,
        setPeriod: PropTypes.func.isRequired,
        setAreaRadius: PropTypes.func.isRequired,
        validateLayer: PropTypes.bool.isRequired,
    };

    constructor(props, context) {
        super(props, context);
        this.state = {
            tab: 'data',
            showBuffer: this.hasBuffer(props.areaRadius),
        };
    }

    componentDidMount() {
        const {
            rows,
            filters,
            defaultPeriod,
            startDate,
            endDate,
            setOrgUnitRoot,
            setPeriod,
        } = this.props;

        const orgUnits = getOrgUnitNodesFromRows(rows);
        const period = getPeriodFromFilters(filters);

        // Set org unit tree root as default
        if (orgUnits.length === 0) {
            setOrgUnitRoot();
        }

        // Set default period from system settings
        if (!period && !startDate && !endDate && defaultPeriod) {
            setPeriod({
                id: defaultPeriod,
            });
        }
    }

    componentDidUpdate(prev) {
        const { areaRadius, validateLayer, onLayerValidation } = this.props;

        if (areaRadius !== prev.areaRadius) {
            this.setState({
                showBuffer: this.hasBuffer(areaRadius),
            });
        }

        if (validateLayer && validateLayer !== prev.validateLayer) {
            onLayerValidation(this.validate());
        }
    }

    render() {
        const {
            // layer options
            areaRadius,
            columns = [],
            endDate,
            eventClustering,
            eventStatus,
            eventCoordinateField,
            eventPointColor,
            eventPointRadius,
            filters = [],
            program,
            programStage,
            rows = [],
            startDate,
        } = this.props;

        const {
            // handlers
            setProgram,
            setProgramStage,
            setEventStatus,
            setEventCoordinateField,
            setEventClustering,
            setEventPointColor,
            setEventPointRadius,
            setUserOrgUnits,
            toggleOrgUnit,
            setPeriod,
            setAreaRadius,
        } = this.props;

        const {
            tab,
            programError,
            programStageError,
            periodError,
            orgUnitsError,
            showBuffer,
        } = this.state;

        const period = getPeriodFromFilters(filters) || {
            id: 'START_END_DATES',
        };

        const selectedUserOrgUnits = getUserOrgUnitsFromRows(rows);

        return (
            <div data-test="eventdialog">
                <Tabs value={tab} onChange={tab => this.setState({ tab })}>
                    <Tab value="data">{i18n.t('Data')}</Tab>
                    <Tab value="period">{i18n.t('Period')}</Tab>
                    <Tab value="orgunits">{i18n.t('Org Units')}</Tab>
                    <Tab value="filter">{i18n.t('Filter')}</Tab>
                    <Tab value="style">{i18n.t('Style')}</Tab>
                </Tabs>
                <div
                    className={styles.tabContent}
                    data-test="eventdialog-content"
                >
                    {tab === 'data' && (
                        <div
                            className={styles.flexRowFlow}
                            data-test="eventdialog-datatab"
                        >
                            <ProgramSelect
                                program={program}
                                onChange={setProgram}
                                className={styles.select}
                                errorText={programError}
                                data-test="eventdialog-programselect"
                            />
                            <ProgramStageSelect
                                program={program}
                                programStage={programStage}
                                onChange={setProgramStage}
                                className={styles.select}
                                errorText={programStageError}
                            />
                            <CoordinateField
                                program={program}
                                programStage={programStage}
                                value={eventCoordinateField}
                                onChange={setEventCoordinateField}
                                className={styles.select}
                            />
                            <EventStatusSelect
                                value={eventStatus}
                                onChange={setEventStatus}
                                className={styles.select}
                            />
                        </div>
                    )}
                    {tab === 'period' && (
                        <div
                            className={styles.flexRowFlow}
                            data-test="eventdialog-periodtab"
                        >
                            <RelativePeriodSelect
                                period={period}
                                startEndDates={true}
                                onChange={setPeriod}
                                className={styles.select}
                            />
                            {period && period.id === 'START_END_DATES' && (
                                <StartEndDates
                                    startDate={startDate}
                                    endDate={endDate}
                                    errorText={periodError}
                                />
                            )}
                        </div>
                    )}
                    {tab === 'orgunits' && (
                        <div
                            className={styles.flexColumnFlow}
                            data-test="eventdialog-orgunittab"
                        >
                            <div className={styles.orgUnitTree}>
                                <OrgUnitTree
                                    selected={getOrgUnitNodesFromRows(rows)}
                                    onClick={toggleOrgUnit}
                                    disabled={
                                        selectedUserOrgUnits.length
                                            ? true
                                            : false
                                    }
                                />
                            </div>
                            <div className={styles.flexColumn}>
                                <UserOrgUnitsSelect
                                    selected={selectedUserOrgUnits}
                                    onChange={setUserOrgUnits}
                                />
                                <SelectedOrgUnits
                                    rows={rows}
                                    units={i18n.t('Events')}
                                    error={orgUnitsError}
                                />
                            </div>
                        </div>
                    )}
                    {tab === 'filter' && (
                        <div
                            className={styles.flexRowFlow}
                            data-test="eventdialog-filtertab"
                        >
                            <FilterGroup
                                program={program}
                                programStage={programStage}
                                filters={columns.filter(
                                    c => c.filter !== undefined
                                )}
                            />
                        </div>
                    )}
                    {tab === 'style' && (
                        <div
                            className={styles.flexColumnFlow}
                            data-test="eventdialog-styletab"
                        >
                            <div className={styles.flexColumn}>
                                <div className={styles.flexInnerColumnFlow}>
                                    <ImageSelect
                                        id="cluster"
                                        img="images/cluster.png"
                                        title={i18n.t('Group events')}
                                        onClick={() => setEventClustering(true)}
                                        isSelected={eventClustering}
                                        className={styles.flexInnerColumn}
                                    />
                                    <ImageSelect
                                        id="nocluster"
                                        img="images/nocluster.png"
                                        title={i18n.t('View all events')}
                                        onClick={() =>
                                            setEventClustering(false)
                                        }
                                        isSelected={!eventClustering}
                                        className={styles.flexInnerColumn}
                                    />
                                </div>
                                <div className={styles.flexInnerColumnFlow}>
                                    <ColorPicker
                                        label={i18n.t('Color')}
                                        color={
                                            cssColor(eventPointColor) ||
                                            EVENT_COLOR
                                        }
                                        onChange={setEventPointColor}
                                        className={styles.flexInnerColumn}
                                    />
                                    <NumberField
                                        label={i18n.t('Radius')}
                                        value={eventPointRadius || EVENT_RADIUS}
                                        onChange={setEventPointRadius}
                                        className={styles.flexInnerColumn}
                                    />
                                </div>
                                <div className={styles.flexInnerColumnFlow}>
                                    <Checkbox
                                        label={i18n.t('Buffer')}
                                        checked={showBuffer}
                                        onChange={this.onShowBufferClick.bind(
                                            this
                                        )}
                                        className={styles.checkboxInline}
                                        disabled={eventClustering}
                                    />
                                    {showBuffer && (
                                        <NumberField
                                            label={i18n.t('Radius in meters')}
                                            value={areaRadius || ''}
                                            onChange={setAreaRadius}
                                            disabled={eventClustering}
                                            className={styles.flexInnerColumn}
                                        />
                                    )}
                                </div>
                            </div>
                            <div className={styles.flexColumn}>
                                {program ? (
                                    <StyleByDataItem />
                                ) : (
                                    <div className={styles.notice}>
                                        <NoticeBox>
                                            {i18n.t(
                                                'You can style events by data element after selecting a program.'
                                            )}
                                        </NoticeBox>
                                    </div>
                                )}
                            </div>
                        </div>
                    )}
                </div>
            </div>
        );
    }

    onShowBufferClick(isChecked) {
        const { setAreaRadius, areaRadius } = this.props;
        setAreaRadius(isChecked ? areaRadius || EVENT_BUFFER : null);
    }

    hasBuffer(areaRadius) {
        return areaRadius !== undefined && areaRadius !== null;
    }

    // TODO: Add to parent class?
    setErrorState(key, message, tab) {
        this.setState({
            [key]: message,
            tab,
        });

        return false;
    }

    validate() {
        const {
            program,
            programStage,
            rows,
            filters,
            startDate,
            endDate,
        } = this.props;

        const period = getPeriodFromFilters(filters) || {
            id: 'START_END_DATES',
        };

        if (!program) {
            return this.setErrorState(
                'programError',
                i18n.t('Program is required'),
                'data'
            );
        }

        if (!programStage) {
            return this.setErrorState(
                'programStageError',
                i18n.t('Program stage is required'),
                'data'
            );
        }

        if (period.id === 'START_END_DATES') {
            const error = getStartEndDateError(startDate, endDate);
            if (error) {
                return this.setErrorState('periodError', error, 'period');
            }
        }

        if (!getOrgUnitsFromRows(rows).length) {
            return this.setErrorState(
                'orgUnitsError',
                i18n.t('No organisation units are selected.'),
                'orgunits'
            );
        }

        return true;
    }
}

export default connect(
    null,
    {
        setProgram,
        setProgramStage,
        setEventStatus,
        setEventCoordinateField,
        setEventClustering,
        setEventPointColor,
        setEventPointRadius,
        setOrgUnitRoot,
        setUserOrgUnits,
        toggleOrgUnit,
        setPeriod,
        setAreaRadius,
    },
    null,
    {
        forwardRef: true,
    }
)(EventDialog);
