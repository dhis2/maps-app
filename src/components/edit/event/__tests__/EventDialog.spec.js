import React from 'react';
import { render } from '@testing-library/react';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import EventDialog from '../EventDialog';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

test('EventDialog displays the dialog with the Data tab open', () => {
    const props = {
        defaultPeriod: 'LAST_12_MONTHS',
        eventClustering: true,
        layer: 'event',
        type: 'Events',
        validateLayer: false,
    };

    const store = {
        orgUnitTree: [],
    };

    const { container } = render(
        <Provider store={mockStore(store)}>
            <EventDialog {...props} onLayerValidation={jest.fn()} />
        </Provider>
    );
    expect(container).toMatchSnapshot();
});
