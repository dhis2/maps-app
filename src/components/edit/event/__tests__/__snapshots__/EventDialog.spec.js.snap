// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventDialog displays the dialog with the Data tab open 1`] = `
<div>
  <div
    data-test="eventdialog"
  >
    <div
      data-test="dhis2-uicore-tabbar"
    >
      <div
        class="jsx-359496394 fixed"
        data-test="dhis2-uicore-tabbar-tabs"
      >
        <button
          class="jsx-3980598635 tab selected"
          data-test="dhis2-uicore-tab"
        >
          <span
            class="jsx-3980598635"
          >
            Data
          </span>
        </button>
        <button
          class="jsx-3980598635 tab"
          data-test="dhis2-uicore-tab"
        >
          <span
            class="jsx-3980598635"
          >
            Period
          </span>
        </button>
        <button
          class="jsx-3980598635 tab"
          data-test="dhis2-uicore-tab"
        >
          <span
            class="jsx-3980598635"
          >
            Org Units
          </span>
        </button>
        <button
          class="jsx-3980598635 tab"
          data-test="dhis2-uicore-tab"
        >
          <span
            class="jsx-3980598635"
          >
            Filter
          </span>
        </button>
        <button
          class="jsx-3980598635 tab"
          data-test="dhis2-uicore-tab"
        >
          <span
            class="jsx-3980598635"
          >
            Style
          </span>
        </button>
      </div>
    </div>
    <div
      class="tabContent"
      data-test="eventdialog-content"
    >
      <div
        class="flexRowFlow"
        data-test="eventdialog-datatab"
      >
        <div
          class="inputField select"
        >
          <div
            class="jsx-1141087211 "
            data-test="dhis2-uiwidgets-singleselectfield"
          >
            <label
              class="jsx-2718078600 "
              data-test="dhis2-uiwidgets-singleselectfield-label"
            >
              <span
                class="jsx-2718078600"
              >
                Program
              </span>
            </label>
            <div
              class="jsx-3111771310 "
              data-test="dhis2-uiwidgets-singleselectfield-content"
            >
              <div
                class="jsx-3415779570 "
                data-test="dhis2-uicore-box"
              >
                <div
                  class="jsx-2510973279 root"
                  data-test="dhis2-uicore-singleselect"
                >
                  <div
                    class="jsx-2510973279 root-input"
                  >
                    <div
                      data-test="dhis2-uicore-select"
                    >
                      <div
                        class="jsx-297747822 root dense"
                        data-test="dhis2-uicore-select-input"
                        tabindex="0"
                      >
                        <div
                          class="jsx-297747822 root-children"
                        >
                          <div
                            class="jsx-2240447017 jsx-4123978364 root"
                          >
                            
                            
                          </div>
                        </div>
                        <div
                          class="jsx-297747822 root-right"
                        >
                          <svg
                            class="jsx-595155045 "
                            version="1.1"
                            viewBox="0 0 16 16"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <g
                              class="jsx-595155045"
                              fill="none"
                              fill-rule="evenodd"
                              id="icon/16/arrow-down"
                              stroke="none"
                              stroke-width="1"
                            >
                              <path
                                class="jsx-595155045"
                                d="M7.29289,11.7071 C7.68342,12.0976 8.31658,12.0976 8.70711,11.7071 L13.7071,6.70711 C14.0976,6.31658 14.0976,5.68342 13.7071,5.29289 C13.3166,4.90237 12.6834,4.90237 12.2929,5.29289 L8,9.58579 L3.70711,5.29289 C3.31658,4.90237 2.68342,4.90237 2.29289,5.29289 C1.90237,5.68342 1.90237,6.31658 2.29289,6.70711 L7.29289,11.7071 Z"
                                fill="#4A5768"
                                id="arrow-down"
                              />
                            </g>
                          </svg>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="inputField select"
        >
          <div
            class="jsx-1141087211 "
            data-test="dhis2-uiwidgets-singleselectfield"
          >
            <label
              class="jsx-2718078600 "
              data-test="dhis2-uiwidgets-singleselectfield-label"
            >
              <span
                class="jsx-2718078600"
              >
                Coordinate field
              </span>
            </label>
            <div
              class="jsx-3111771310 "
              data-test="dhis2-uiwidgets-singleselectfield-content"
            >
              <div
                class="jsx-3415779570 "
                data-test="dhis2-uicore-box"
              >
                <div
                  class="jsx-2510973279 root"
                  data-test="dhis2-uicore-singleselect"
                >
                  <div
                    class="jsx-2510973279 root-input"
                  >
                    <div
                      data-test="dhis2-uicore-select"
                    >
                      <div
                        class="jsx-297747822 root dense"
                        data-test="dhis2-uicore-select-input"
                        tabindex="0"
                      >
                        <div
                          class="jsx-297747822 root-children"
                        >
                          <div
                            class="jsx-2240447017 jsx-4123978364 root"
                          >
                            <div
                              class="jsx-2240447017 jsx-4123978364 root-input"
                            >
                              <div
                                class="jsx-3691749911 root"
                              >
                                Event location
                              </div>
                            </div>
                          </div>
                        </div>
                        <div
                          class="jsx-297747822 root-right"
                        >
                          <svg
                            class="jsx-595155045 "
                            version="1.1"
                            viewBox="0 0 16 16"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <g
                              class="jsx-595155045"
                              fill="none"
                              fill-rule="evenodd"
                              id="icon/16/arrow-down"
                              stroke="none"
                              stroke-width="1"
                            >
                              <path
                                class="jsx-595155045"
                                d="M7.29289,11.7071 C7.68342,12.0976 8.31658,12.0976 8.70711,11.7071 L13.7071,6.70711 C14.0976,6.31658 14.0976,5.68342 13.7071,5.29289 C13.3166,4.90237 12.6834,4.90237 12.2929,5.29289 L8,9.58579 L3.70711,5.29289 C3.31658,4.90237 2.68342,4.90237 2.29289,5.29289 C1.90237,5.68342 1.90237,6.31658 2.29289,6.70711 L7.29289,11.7071 Z"
                                fill="#4A5768"
                                id="arrow-down"
                              />
                            </g>
                          </svg>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="inputField select"
        >
          <div
            class="jsx-1141087211 "
            data-test="dhis2-uiwidgets-singleselectfield"
          >
            <label
              class="jsx-2718078600 "
              data-test="dhis2-uiwidgets-singleselectfield-label"
            >
              <span
                class="jsx-2718078600"
              >
                Event status
              </span>
            </label>
            <div
              class="jsx-3111771310 "
              data-test="dhis2-uiwidgets-singleselectfield-content"
            >
              <div
                class="jsx-3415779570 "
                data-test="dhis2-uicore-box"
              >
                <div
                  class="jsx-2510973279 root"
                  data-test="dhis2-uicore-singleselect"
                >
                  <div
                    class="jsx-2510973279 root-input"
                  >
                    <div
                      data-test="dhis2-uicore-select"
                    >
                      <div
                        class="jsx-297747822 root dense"
                        data-test="dhis2-uicore-select-input"
                        tabindex="0"
                      >
                        <div
                          class="jsx-297747822 root-children"
                        >
                          <div
                            class="jsx-2240447017 jsx-4123978364 root"
                          >
                            <div
                              class="jsx-2240447017 jsx-4123978364 root-input"
                            >
                              <div
                                class="jsx-3691749911 root"
                              >
                                All
                              </div>
                            </div>
                          </div>
                        </div>
                        <div
                          class="jsx-297747822 root-right"
                        >
                          <svg
                            class="jsx-595155045 "
                            version="1.1"
                            viewBox="0 0 16 16"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <g
                              class="jsx-595155045"
                              fill="none"
                              fill-rule="evenodd"
                              id="icon/16/arrow-down"
                              stroke="none"
                              stroke-width="1"
                            >
                              <path
                                class="jsx-595155045"
                                d="M7.29289,11.7071 C7.68342,12.0976 8.31658,12.0976 8.70711,11.7071 L13.7071,6.70711 C14.0976,6.31658 14.0976,5.68342 13.7071,5.29289 C13.3166,4.90237 12.6834,4.90237 12.2929,5.29289 L8,9.58579 L3.70711,5.29289 C3.31658,4.90237 2.68342,4.90237 2.29289,5.29289 C1.90237,5.68342 1.90237,6.31658 2.29289,6.70711 L7.29289,11.7071 Z"
                                fill="#4A5768"
                                id="arrow-down"
                              />
                            </g>
                          </svg>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
